Add utility for opening the config file

neithernut opened the pull request:
This basically tries to load the config file.
One bit is still missing (waiting for #151 -- the JSON parser).
Targets #220.

Commit:
Add initial interface of config loading utility

Commit:
Add function for testing variants

Commit:
Add convenience function for open()ing the config file

Commit:
Add implementation of ws_config_load()

matthiasbeyer commented on src/util/config_file.c:
```diff
@@ -0,0 +1,195 @@
+        return fd;
+    }
+
+    //!< @todo fill in 2nd parameter
```
What 2nd parameter?

matthiasbeyer commented on src/util/config_file.c:
```diff
@@ -0,0 +1,195 @@
+        return -1;
+    }
+
+    return ws_command_processor_new(fd, deserializer, NULL)? 0 : -1;
```
Please add a space right before the `?` operator.

matthiasbeyer commented on src/util/config_file.c:
```diff
@@ -0,0 +1,195 @@
+    {
+        char const* home = getenv(HOME);
+        size_t homelen = strlen(home);
+        char buf[homelen + 10];
```
This `10` here appears to be a magic constant to me, so please document it.

matthiasbeyer commented on src/util/config_file.c:
```diff
@@ -0,0 +1,195 @@
+    {
+        // setup all the buffers and temporaries
+        char const* xdg_dirs = getenv(XDG_CONFIG_DIRS);
+        char buf[strlen(xdg_dirs)+1];
```
please add spaces around the `+` operator.

matthiasbeyer commented:
Looks good besides the three remaining comments.

Commit:
Read config file from main()

neithernut commented:
I pushed some changes. The code should now be complete. But merging it having an _totally_ incomplete deserializer.

neithernut commented:
Needs a little bit more work.

---

Disregard that. It's ready

neithernut commented:
Testing. Alread found a few errors in the program flow (and fixed them).
I think the config loading function is ready now, but it seems the pipeline is still not working. I'll notify you as soon as I tracked down the problems.

matthiasbeyer commented:
ACK.

Commit:
Make connector return EOF if it hits the end of file

Commit:
Rewrite error detection logic in command processor

neithernut commented:
I encountered a bug in `ws_deserialize()` (it read an entire message but didn't return it). Investigating further. @matthiasbeyer, could you have a look at it? I put the following content in a file `waysome.json` in `$PWD`:

```
{ "UID": "my awesome uid", "commands": [ { "add": [1, 2] } ] }


```

neithernut commented:
Turned out the user documentation was lying about the format.

Commit:
Fix argument preparation.

First of all, we forgot `break`s in an switch-case.
Secondly, we reversed the order the arguments are pushed, so the
arguments showing up first end up lower on the stack.

Commit:
Fix inverse logic bug in command subsystem macros

Commit:
Fix ws_processor_stack_value_at()

Commit:
Fix constructor of value_reply

Commit:
Fix constructor of error_reply

neithernut commented:
There were a few quite embarrasing bugs, but it looks like it all works now.

matthiasbeyer commented:
Looks good to me. ACK.
